src/runtime/hexagon_remote/remote.cpp

```diff
@@ -44,15 +95,239 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+        qurt_thread_attr_set_priority(&thread_attr, 100);
+        qurt_thread_create(&threads[i], &thread_attr, thread_server, (void *)i);
+    }
+    FARF(LOW, "HVX_TP: Master Thread: Created threads\n");
+}
+void qurt_thread_pool_init() {
+    FARF(LOW, "HVX_TP: Master Thread: Initializing the thread pool\n");
+    qurt_mutex_init(&(work_queue.work_mutex));
+    qurt_sem_init_val(&wait_for_work, 0);
+
+    initialize_work_queue();
 
+    create_threads(NUM_WORKER_THREADS_TO_CREATE);
+
+    thread_pool_initialized = true;
+    FARF(LOW, "HVX_TP: Master Thread: Thread pool initialized\n");
+}
+int find_empty_job() {
+    int i;
+    work *jobs = work_queue.jobs;
+    for (i = 0; i < MAX_WORKLOADS; ++i) {
+        if (jobs[i].done()) {
+            return i;
+        }
+    }
+    return -1;
+}
 int halide_do_par_for(void *user_context, halide_task_t f,
                       int min, int size, uint8_t *closure) {
-    for (int x = min; x < min + size; x++) {
-        int result = halide_do_task(user_context, f, x, closure);
-        if (result) {
-            return result;
+    FARF(LOW, "HVX_TP: Master Thread: halide_do_par_for\n");
+
+    // 1. If the thread pool hasn't been initialized, initiliaze it.
+    // This involves.
+    //    a) Creating the threads.
+    //    b) Acquiring a lock on the work queue and clearing the jobs therein
+    //    c) Setting up a semaphore on which worker threads sleep until awoken
+    //       by this thread i.e. the master thread.
+    if (!thread_pool_initialized) {
+        qurt_thread_pool_init();
+    }
+
+    // 2. Lock the work queue again.
+    lock_work_queue();
+
+    // 3. Put work in the global work queue.
+    int next_workload = work_queue.next_workload;
+
+    if (!work_queue.jobs[next_workload].done()) {
+        FARF(LOW, "HVX_TP: Master Thread: Job at idx = %d is still not done, so looking for an empty job\n", next_workload);
+        next_workload = find_empty_job();
+        if (next_workload == -1) {
+            // We should really just be going to sleep here, until woken up by another thread.
+            FARF(LOW, "HVX_TP: Master Thread: Work queue is full\n");
+            return -1;
         }
+        FARF(LOW, "HVX_TP: Master Thread: Found empty job at idx  = %d\n", next_workload);
     }
+
+    FARF(LOW, "HVX_TP: Master Thread: Putting job in the work queue at idx = %d, min = %d, size = %d, user_context = 0x%x, f= 0x%x, \n",
+            next_workload, min, size, user_context, f);
+
+    work *job = &work_queue.jobs[next_workload];
+    job->f = f;
+    job->user_context = user_context;
+    job->next = min;          // Start at this index.
+    job->max  = min + size;   // Keep going until one less than this index.
+    job->closure = closure;   // Use this closure.
+    job->exit_status = 0;     // The job hasn't failed yet
+    job->active_workers = 0;  // Nobody is working on this yet
+    // We are implementing a circular array to store jobs.
```

Comment:
Seems like this needs a check to make sure the circular buffer head isn't catching up to the tail. To check my understanding, this would only happen in some crazy nested parallelism use case, correct?